src/sources/docker.rs

```diff
@@ -107,15 +94,31 @@ impl SourceConfig for DockerConfig {
         _name: &str,
         _globals: &GlobalOptions,
         shutdown: ShutdownSignal,
-        out: Sender<Event>,
+        out: mpsc01::Sender<Event>,
     ) -> crate::Result<super::Source> {
-        DockerSource::new(
-            self.clone().with_empty_partial_event_marker_field_as_none(),
-            out,
-            shutdown,
-        )
-        .map(Box::new)
-        .map(|source| source as Box<_>)
+        let source = DockerSource::new(self.clone(), out, shutdown.clone())?;
+
+        // Capture currently running containers, and do main future(run)
+        let fut = async move {
+            match source.handle_running_containers().await {
+                Ok(source) => source.run().await,
+                Err(error) => {
+                    error!(message = "Listing currently running containers, failed", %error);
+                }
+            }
+        };
+
+        // Once this ShutdownSignal resolves it will drop DockerSource and by extension it's ShutdownSignal.
+        Ok(Box::new(
+            async move {
+                Ok(tokio::select! {
+                    _ = fut => {}
+                    _ = shutdown.compat() => {}
+                })
```

Comment:
Not sure what's better `tokio::select!` or `futures::select!`?

Reply:
Yeap, `tokio` more about runtime. I wondered is there any other difference? For example `mpsc` faster in `tokio` than in `futures`. Also we already use `tokio::select!`: https://github.com/timberio/vector/blob/01b37785f38159388636ff9954a3675c1ae0d049/src/sinks/file/mod.rs#L118

Reply:
Not that I know of. It should be a small difference in both cases. With what's said here, retaining `tokio::select` seems fine.